src/hotspot/share/cds/filemap.cpp

```diff
@@ -2678,11 +2681,60 @@ ClassFileStream* FileMapInfo::open_stream_for_jvmti(InstanceKlass* ik, Handle cl
   const char* const file_name = ClassLoader::file_name_for_class_name(class_name,
                                                                       name->utf8_length());
   ClassLoaderData* loader_data = ClassLoaderData::class_loader_data(class_loader());
-  ClassFileStream* cfs = cpe->open_stream_for_loader(THREAD, file_name, loader_data);
+  ClassFileStream* cfs;
+  if (class_loader() != nullptr && !cpe->is_modules_image()) {
+    cfs = get_stream_from_class_loader(class_loader, cpe, file_name, CHECK_NULL);
+  } else {
+    cfs = cpe->open_stream_for_loader(THREAD, file_name, loader_data);
+  }
   assert(cfs != nullptr, "must be able to read the classfile data of shared classes for built-in loaders.");
   log_debug(cds, jvmti)("classfile data for %s [%d: %s] = %d bytes", class_name, path_index,
                         cfs->source(), cfs->length());
   return cfs;
 }
 
+ClassFileStream* FileMapInfo::get_stream_from_class_loader(Handle class_loader,
+                                                           ClassPathEntry* cpe,
+                                                           const char* file_name,
+                                                           TRAPS) {
+  JavaValue result(T_OBJECT);
+  TempNewSymbol class_name_sym = SymbolTable::new_symbol(file_name);
+  Handle ext_class_name = java_lang_String::externalize_classname(class_name_sym, CHECK_NULL);
+
+  // InputStream ClassLoader.getResourceAsStream(String name)
+  JavaCalls::call_virtual(&result,
```

Comment:
Do we really have to make two Java upcalls to do this? What is the performance hit? Can we declare a helper method in ClassLoader that does it in one? Is there any risk of getting into a recursive loop here?

Reply:
I've added the following method in ClassLoader so we only make one Java upcalls:
`private byte[] getResourceAsByteArray(String name)`
This is a special case of using CDS and CFLH. So I think it can take a bit of performance hit. Better than experiencing a VM crash.
Since the new function doesn't have any global state, it should be able to handle recursive call.